### PR TITLE
Added examples and improved docstrings

### PR DESCRIPTION
**Pull request by ZiadAmerr**
**Changes**
- Removed code fragments by passing a `kwarg` variable containing the duplicate fragments
- Added docstring to helper functions missing docstrings
- Improved current docstring
- Added examples to some helper functions

**TODOs** either:
- mark something that should be reviewed by @CatB1t 
- mark something I am suspicious that it might be a bug
- unfinished code part

### CODE REVIEW

**Comment by ZiadAmerr** [comment]
This PR: #13324 is also working on the same task but on different parts of the code.

File: ivy_tests/test_ivy/helpers/hypothesis_helpers/array_helpers.py [hunk 1]
```diff
@@ -925,6 +991,23 @@ def array_values(
     Returns
     -------
         A strategy that draws a list.
+
+    Examples
+    --------
+    >>> array_values(
+    ...     dtype=get_dtypes("float"),
+    ...     shape=st.shared(get_shape()),
```
Comment (CatB1t) [inline ivy_tests/test_ivy/helpers/hypothesis_helpers/array_helpers.py:999]: Let's eliminate the use of `st.shared` here not to cause confusion, also let's change the `get_dtypes` to have `valid` instead of `float`.

File: ivy_tests/test_ivy/helpers/hypothesis_helpers/dtype_helpers.py [hunk 2]
```diff
@@ -246,7 +277,26 @@ def array_dtypes(
...
+        A strategy that draws a list of data types.
+
+    Examples
+    --------
+    >>> array_dtypes(
+    ...     available_dtypes=get_dtypes("numeric"),
+    ...     shared_dtype=True,
+    ... )
+    ['float64']
+    ['int8', 'int8']
```
Comment (CatB1t) [inline ivy_tests/test_ivy/helpers/hypothesis_helpers/dtype_helpers.py:289]: Let's only provide one output of the single function call, so not to cause confusion.

File: ivy_tests/test_ivy/helpers/hypothesis_helpers/dtype_helpers.py [hunk 3]
```diff
@@ -304,7 +354,7 @@ def get_castable_dtype(draw, available_dtypes, dtype: str, x: Optional[list] = N
     Returns
     -------
     ret
-        A tuple of inputs and castable dtype.
+        A strategy that draws a tuple of inputs and a castable dtype.
```
Comment (CatB1t) [inline ivy_tests/test_ivy/helpers/hypothesis_helpers/dtype_helpers.py:357]: The return seems to be the result of `draw` function, which doesn't return a strategy, I think the previous description is more suitable.

File: ivy_tests/test_ivy/helpers/hypothesis_helpers/general_helpers.py [hunk 4]
```diff
@@ -239,17 +233,18 @@ def get_mean_std(draw, *, dtype):
 
     Returns
     -------
-    A strategy that can be used in the @given hypothesis decorator.
+        A strategy that can be used in the @given hypothesis decorator.
     """
-    none_or_float = none_or_float = number_helpers.floats(dtype=dtype) | st.none()
+    # TODO: Does this actually work?
+    none_or_float = number_helpers.floats(dtype=dtype) | st.none()
```
Comment (CatB1t) [inline ivy_tests/test_ivy/helpers/hypothesis_helpers/general_helpers.py:239]: Yes, this is the same as writing `st.one_of(number_helpers.floats(dtype=dtype), st.none())`

File: ivy_tests/test_ivy/helpers/hypothesis_helpers/general_helpers.py [hunk 5]
```diff
@@ -401,6 +396,24 @@ def sort_key(ele, max_len):
 
 @st.composite
 def x_and_filters(draw, dim: int = 2, transpose: bool = False, depthwise=False):
+    """Draws a random x and filters for a convolution.  # TODO: What should be done here?
```
Comment (CatB1t) [inline ivy_tests/test_ivy/helpers/hypothesis_helpers/general_helpers.py:399]: Can you elaborate more on what you mean by the TODO comment?

File: ivy_tests/test_ivy/test_frontends/test_torch/test_loss_functions.py [hunk 6]
```diff
@@ -443,7 +443,7 @@ def test_torch_huber_loss(
     pred_dtype, pred = input_dtype[0], x[0]
     true_dtype, true = input_dtype[1], x[1]
     helpers.test_frontend_function(
-        input_dtypes=[pred_dtype, true_dtype],
+        input_dtypes=input_dtype,
```
Comment (CatB1t) [inline ivy_tests/test_ivy/test_frontends/test_torch/test_loss_functions.py:446]: Let's revert this unrelated change to make the PR clean.

**Review by CatB1t — CHANGES_REQUESTED** [review]
Hello @ZiadAmerr great work!, I've left few comments then we can get this PR merged later after the requested changes, just to not cause confusion, let's only provide one output for the single function call in **all** examples, we should also add a note that this is output is completely _randomly_ generated.

for the TODOs questions, I'd suggest you use git blame to find out the author of the code, as they would be a better candidate to answer most of these questions than me. Thanks!

**Comment by ZiadAmerr** [comment]
Hi @CatB1t, thanks for the review; but don't you think providing multiple outputs for a single function call would really show what the randomization really does? Showing one output only would show a single case while we really have multiple cases, and showing some of them will be better for the reader so that they can better understand the variations of that function's output?

**Comment by CatB1t** [comment]
I agree with that, but let's add the multiple function calls too, so it's not misleading that the function outputs multiple values for a single call.

File: ivy_tests/test_ivy/helpers/hypothesis_helpers/array_helpers.py [hunk 7]
```diff
@@ -787,7 +1021,7 @@ def array_indices_axis(
 def arrays_and_axes(
     draw,
     available_dtypes=get_dtypes("float"),
-    allow_none=False,
+    allow_none=False,  # TODO: Should this exist? - 1
```
Comment (CatB1t) [inline ivy_tests/test_ivy/helpers/hypothesis_helpers/array_helpers.py:1024]: Let's get this comment removed for merging.

File: ivy_tests/test_ivy/helpers/hypothesis_helpers/array_helpers.py [hunk 8]
```diff
@@ -796,11 +1030,107 @@ def arrays_and_axes(
...
+    ... )
+    ([array([-6.72e-05, -6.72e-05, -6.72e-05, -6.72e-05, -6.72e-05],
+        dtype=float16)], 0)
+    """
     shapes = list()
     for _ in range(num):
         shape = draw(
             gh.get_shape(
-                allow_none=False,
+                allow_none=False,  # TODO: Should this exist? - 1
```
Comment (CatB1t) [inline ivy_tests/test_ivy/helpers/hypothesis_helpers/array_helpers.py:1133]: Let's get this comment removed for merging.

File: ivy_tests/test_ivy/helpers/hypothesis_helpers/array_helpers.py [hunk 9]
```diff
@@ -820,8 +1150,8 @@ def arrays_and_axes(
             draw(array_values(dtype=dtype[0], shape=shape, min_value=-20, max_value=20))
         )
     if force_int_axis:
-        if len(shape) <= 2:
-            axes = draw(st.one_of(st.integers(0, len(shape) - 1), st.none()))
+        if len(shape) <= 2:  # TODO: What is "shape"? It doesn't seem to be defined previously - 2
```
Comment (CatB1t) [inline ivy_tests/test_ivy/helpers/hypothesis_helpers/array_helpers.py:1153]: Let's get this comment removed for merging.

File: ivy_tests/test_ivy/helpers/hypothesis_helpers/array_helpers.py [hunk 10]
```diff
@@ -820,8 +1150,8 @@ def arrays_and_axes(
...
+            axes = draw(st.one_of(st.integers(0, len(shape) - 1), st.none()))  # TODO: st.none() might overlap with allow_none? - 1
```
Comment (CatB1t) [inline ivy_tests/test_ivy/helpers/hypothesis_helpers/array_helpers.py:1154]: Let's get this comment removed for merging.

File: ivy_tests/test_ivy/helpers/hypothesis_helpers/dtype_helpers.py [hunk 11]
```diff
@@ -99,11 +99,53 @@ def get_dtypes(
         returns the complete list of valid types
     none
         allow none in the list of valid types
+    key
+        TODO: add description
```
Comment (CatB1t) [inline ivy_tests/test_ivy/helpers/hypothesis_helpers/dtype_helpers.py:103]: The key parameter is here for `st.shared`, if provided, a shared value will be drawn using `st.shared`

File: ivy_tests/test_ivy/helpers/hypothesis_helpers/dtype_helpers.py [hunk 12]
```diff
@@ -99,11 +99,53 @@ def get_dtypes(
...
+    prune_function
+        TODO: add description
```
Comment (CatB1t) [inline ivy_tests/test_ivy/helpers/hypothesis_helpers/dtype_helpers.py:105]: if `prune_function` is True, the drawn data types will be specific for the currently running test.

File: ivy_tests/test_ivy/helpers/hypothesis_helpers/dtype_helpers.py [hunk 13]
```diff
@@ -99,11 +99,53 @@ def get_dtypes(
...
+    ... )
+    ['uint16']
+
+    >>> get_dtypes(
+    ...     kind='numeric',
+    ...     full=False,
+    ... )
+    ['complex64']
+
+    # TODO: Needs an example that utilizes key and prune_function
```
Comment (CatB1t) [inline ivy_tests/test_ivy/helpers/hypothesis_helpers/dtype_helpers.py:148]: Let's get the examples added.

File: ivy_tests/test_ivy/helpers/hypothesis_helpers/general_helpers.py [hunk 14]
```diff
@@ -239,17 +233,18 @@ def get_mean_std(draw, *, dtype):
 
     Returns
     -------
-    A strategy that can be used in the @given hypothesis decorator.
+        A strategy that can be used in the @given hypothesis decorator.
     """
-    none_or_float = none_or_float = number_helpers.floats(dtype=dtype) | st.none()
+    # TODO: Does this actually work?
```
Comment (CatB1t) [inline ivy_tests/test_ivy/helpers/hypothesis_helpers/general_helpers.py:238]: Let's remove the comment.

File: ivy_tests/test_ivy/helpers/hypothesis_helpers/general_helpers.py [hunk 15]
```diff
@@ -288,14 +283,14 @@ def get_axis(
     shape,
     allow_neg=True,
     allow_none=False,
-    sorted=True,
+    sorted=True,  # TODO: should be renamed to sort_values
```
Comment (CatB1t) [inline ivy_tests/test_ivy/helpers/hypothesis_helpers/general_helpers.py:286]: Let's get it renamed before the merge.

File: ivy_tests/test_ivy/helpers/hypothesis_helpers/general_helpers.py [hunk 16]
```diff
@@ -443,51 +456,64 @@ def x_and_filters(draw, dim: int = 2, transpose: bool = False, depthwise=False):
...
-            output_shape,
-        )
-    return dtype, vals, filters, dilations, data_format, strides, padding
+        return *ret, output_shape
+    return ret
 
 
 @st.composite
 def embedding_helper(draw):
+    """Helper function for embedding layer tests. # TODO: What does this function do?
```
Comment (CatB1t) [inline ivy_tests/test_ivy/helpers/hypothesis_helpers/general_helpers.py:494]: Let's reach out for the owner of this code, and add the proper description.

**Review by CatB1t — CHANGES_REQUESTED** [review]

File: ivy_tests/test_ivy/helpers/hypothesis_helpers/array_helpers.py [hunk 17]
```diff
@@ -1198,7 +1545,7 @@ def arrays_for_pooling(
         )
     if array_dim == 3:
         kernel = draw(st.tuples(st.integers(1, in_shape[1])))
-    new_kernel = kernel
+    new_kernel = kernel  # TODO: This will raise an error if array_dim not in [5, 4, 3]
```
Comment (ZiadAmerr) [inline ivy_tests/test_ivy/helpers/hypothesis_helpers/array_helpers.py:1548]: Hi @CatB1t, can you take a look at this?

Comment (CatB1t) [reply ivy_tests/test_ivy/helpers/hypothesis_helpers/array_helpers.py:1548]: Hello @ZiadAmerr , it's better to reach out to the author of the code through git blame, meanwhile, we can merge it if it's ready as it doesn't seem to be a docs-related issue.

File: ivy_tests/test_ivy/helpers/hypothesis_helpers/array_helpers.py [hunk 18]
```diff
@@ -1583,7 +1583,71 @@ def arrays_for_pooling(
...
+        x[0] = np.transpose(x[0], (0, dim - 1, *range(1, dim - 1)))
+
+    ret = dtype, x, kernel, strides, padding
     if return_dilation:
         return *ret, dilations
     return ret
+
+
+@st.composite
+def dtype_array_index(
```
Comment (CatB1t) [inline ivy_tests/test_ivy/helpers/hypothesis_helpers/array_helpers.py]: Can we revert adding this strategy, and make it in a seperate PR?

File: ivy_tests/test_ivy/helpers/hypothesis_helpers/array_helpers.py [hunk 19]
```diff
@@ -1583,7 +1583,71 @@ def arrays_for_pooling(
         padding = draw(st.sampled_from(["VALID", "SAME"]))
 
     strides = draw(st.tuples(st.integers(1, min(kernel))))
-    ret = [dtype, x, kernel, strides, padding]
+
```
Comment (CatB1t) [inline ivy_tests/test_ivy/helpers/hypothesis_helpers/array_helpers.py:1586]: I'd also suggest we do this one in a seperate PR, to make things cleaner, this PR should only contain improving docstring.

**Review by CatB1t — CHANGES_REQUESTED** [review]

**Review by CatB1t — CHANGES_REQUESTED** [review]
Please ignore the recent review, seems like GitHub wasn't showing proper changes, anyways can we get the conflicts resolved so I can merge it? great work Ziad! Thanks 😄

**Comment by ZiadAmerr** [comment]
@CatB1t done :)

File: ivy_tests/test_ivy/test_frontends/test_numpy/helpers.py [hunk 20]
```diff
@@ -188,7 +188,38 @@ def handle_where_and_array_bools(where, input_dtype, test_flags):
     return where, input_dtype, test_flags
 
 
-# Casting helper
+def handle_dtype_and_casting(
```
Comment (CatB1t) [inline ivy_tests/test_ivy/test_frontends/test_numpy/helpers.py:191]: Can we revert this change here too? Thanks :)

**Review by CatB1t — CHANGES_REQUESTED** [review]
Just a small change then i'll merge it, thanks @ZiadAmerr !

File: ivy_tests/test_ivy/test_frontends/test_numpy/helpers.py (hunk 20 above)
Comment (ZiadAmerr) [reply ivy_tests/test_ivy/test_frontends/test_numpy/helpers.py:191]: This was a resolved conflict not new code I didn't write this

File: ivy_tests/test_ivy/test_frontends/test_numpy/helpers.py [hunk 21]
```diff
@@ -188,7 +188,38 @@ def handle_where_and_array_bools(where, input_dtype, test_flags):
...
+            dtype = helpers.get_dtypes(
+                get_dtypes_kind,
+                index=get_dtypes_index,
+                full=False,
+                none=get_dtypes_none,
+                key=get_dtypes_key,
+            )
+    return dtype, dtypes, casting
+
+
```
Comment (ZiadAmerr) [inline ivy_tests/test_ivy/test_frontends/test_numpy/helpers.py:222]: Removing it from here :)
```suggestion
```

File: ivy_tests/test_ivy/test_frontends/test_numpy/helpers.py [hunk 22]
```diff
@@ -188,7 +188,6 @@ def handle_where_and_array_bools(where, input_dtype, test_flags):
     return where, input_dtype, test_flags
```
Comment (ZiadAmerr) [inline ivy_tests/test_ivy/test_frontends/test_numpy/helpers.py:190]:
```suggestion

# Casting helper
```

File: ivy_tests/test_ivy/helpers/hypothesis_helpers/number_helpers.py [hunk 23]
```diff
@@ -106,26 +108,25 @@ def floats(
...
-                max_value=float_of(max_value, floats_info[dtype]["width"]),
-                allow_nan=allow_nan,
-                allow_subnormal=allow_subnormal,
-                allow_infinity=allow_inf,
-                width=floats_info[dtype]["width"],
-                exclude_min=exclude_min,
-                exclude_max=exclude_max,
+                min_value=float_of(abs_smallest_val, width),
+                max_value=float_of(max_value, width),
+                **float_strategy_kwargs
```
Comment (ZiadAmerr) [inline ivy_tests/test_ivy/helpers/hypothesis_helpers/number_helpers.py:129]:
```suggestion
                min_value=float_of(abs_smallest_val, floats_info[dtype]["width"]),
                max_value=float_of(max_value, floats_info[dtype]["width"]),
                allow_nan=allow_nan,
                allow_subnormal=allow_subnormal,
                allow_infinity=allow_inf,
                width=floats_info[dtype]["width"],
                exclude_min=exclude_min,
                exclude_max=exclude_max,
```

File: ivy_tests/test_ivy/helpers/hypothesis_helpers/number_helpers.py [hunk 24]
```diff
@@ -106,26 +108,25 @@ def floats(
...
-                max_value=float_of(-abs_smallest_val, floats_info[dtype]["width"]),
-                allow_nan=allow_nan,
-                allow_subnormal=allow_subnormal,
-                allow_infinity=allow_inf,
-                width=floats_info[dtype]["width"],
-                exclude_min=exclude_min,
-                exclude_max=exclude_max,
+                min_value=float_of(min_value, width),
+                max_value=float_of(-abs_smallest_val, width),
+                **float_strategy_kwargs
```
Comment (ZiadAmerr) [inline ivy_tests/test_ivy/helpers/hypothesis_helpers/number_helpers.py:124]:
```suggestion
                min_value=float_of(min_value, floats_info[dtype]["width"]),
                max_value=float_of(-abs_smallest_val, floats_info[dtype]["width"]),
                allow_nan=allow_nan,
                allow_subnormal=allow_subnormal,
                allow_infinity=allow_inf,
                width=floats_info[dtype]["width"],
                exclude_min=exclude_min,
                exclude_max=exclude_max,
```

File: ivy_tests/test_ivy/helpers/hypothesis_helpers/number_helpers.py [hunk 25]
```diff
@@ -106,26 +108,25 @@ def floats(
...
     else:
+        width = floats_info[dtype]["width"]
+        float_strategy_kwargs = {
+            "allow_nan": allow_nan,
+            "allow_subnormal": allow_subnormal,
+            "allow_infinity": allow_inf,
+            "width": width,
+            "exclude_min": exclude_min,
+            "exclude_max": exclude_max,
+        }
```
Comment (ZiadAmerr) [inline ivy_tests/test_ivy/helpers/hypothesis_helpers/number_helpers.py:119]:
```suggestion
```

File: ivy_tests/test_ivy/helpers/hypothesis_helpers/general_helpers.py [hunk 26]
```diff
@@ -450,22 +465,23 @@ def x_and_filters(draw, dim: int = 2, transpose: bool = False, depthwise=False):
         x_shape = (batch_size,) + x_dim + (input_channels,)
     else:
         x_shape = (batch_size, input_channels) + x_dim
+
+    common_kwargs = {
+        "dtype": dtype[0],
+        "large_abs_safety_factor": 3,
+        "small_abs_safety_factor": 4,
+        "safety_factor_scale": "log",
+    }
```
Comment (ZiadAmerr) [inline ivy_tests/test_ivy/helpers/hypothesis_helpers/general_helpers.py:474]:
```suggestion
```

File: ivy_tests/test_ivy/helpers/hypothesis_helpers/general_helpers.py [hunk 27]
```diff
@@ -450,22 +465,23 @@ def x_and_filters(draw, dim: int = 2, transpose: bool = False, depthwise=False):
...
     vals = draw(
         array_helpers.array_values(
-            dtype=dtype[0],
             shape=x_shape,
-            large_abs_safety_factor=3,
-            small_abs_safety_factor=4,
-            safety_factor_scale="log",
+            **common_kwargs,
```
Comment (ZiadAmerr) [inline ivy_tests/test_ivy/helpers/hypothesis_helpers/general_helpers.py:478]:
```suggestion
            dtype=dtype[0],
            large_abs_safety_factor=3,
            small_abs_safety_factor=4,
            safety_factor_scale="log",
```

File: ivy_tests/test_ivy/helpers/hypothesis_helpers/general_helpers.py [hunk 28]
```diff
@@ -450,22 +465,23 @@ def x_and_filters(draw, dim: int = 2, transpose: bool = False, depthwise=False):
...
         )
     )
     filters = draw(
         array_helpers.array_values(
-            dtype=dtype[0],
             shape=filter_shape,
-            large_abs_safety_factor=3,
-            small_abs_safety_factor=4,
-            safety_factor_scale="log",
+            **common_kwargs
```
Comment (ZiadAmerr) [inline ivy_tests/test_ivy/helpers/hypothesis_helpers/general_helpers.py:484]:
```suggestion
            dtype=dtype[0],
            large_abs_safety_factor=3,
            small_abs_safety_factor=4,
            safety_factor_scale="log",
```

File: ivy_tests/test_ivy/helpers/hypothesis_helpers/general_helpers.py [hunk 29]
```diff
@@ -204,25 +210,17 @@ def get_shape(
...
+        shape = draw(st.none() | lists_strategy)
     else:
-        shape = draw(
-            st.lists(
-                number_helpers.ints(min_value=min_dim_size, max_value=max_dim_size),
-                min_size=min_num_dims,
-                max_size=max_num_dims,
-            )
-        )
+        shape = draw(lists_strategy)
```
Comment (ZiadAmerr) [inline ivy_tests/test_ivy/helpers/hypothesis_helpers/general_helpers.py:223]:
```suggestion
        shape = draw(
            st.lists(
                number_helpers.ints(min_value=min_dim_size, max_value=max_dim_size),
                min_size=min_num_dims,
                max_size=max_num_dims,
            )
        )
```

File: ivy_tests/test_ivy/helpers/hypothesis_helpers/general_helpers.py [hunk 30]
```diff
@@ -204,25 +210,17 @@ def get_shape(
...
     if allow_none:
-        shape = draw(
-            st.none()
-            | st.lists(
-                number_helpers.ints(min_value=min_dim_size, max_value=max_dim_size),
-                min_size=min_num_dims,
-                max_size=max_num_dims,
-            )
-        )
+        shape = draw(st.none() | lists_strategy)
```
Comment (ZiadAmerr) [inline ivy_tests/test_ivy/helpers/hypothesis_helpers/general_helpers.py:221]:
```suggestion
        shape = draw(
            st.none()
            | st.lists(
                number_helpers.ints(min_value=min_dim_size, max_value=max_dim_size),
                min_size=min_num_dims,
                max_size=max_num_dims,
            )
        )
```
